pyteal/ast/abi/type.py

```diff
@@ -221,17 +220,23 @@ def store_into(self, output: BaseType) -> Expr:
                 f"expected type_spec {self.produced_type_spec()} but get {output.type_spec()}"
             )
 
-        declaration = self.computation.subroutine.get_declaration()
-
-        if declaration.deferred_expr is None:
-            raise TealInputError(
-                "ABI return subroutine must have deferred_expr to be not-None."
-            )
-        if declaration.deferred_expr.type_of() != output.type_spec().storage_type():
-            raise TealInputError(
-                f"ABI return subroutine deferred_expr is expected to be typed {output.type_spec().storage_type()}, "
-                f"but has type {declaration.deferred_expr.type_of()}."
-            )
+        # HANG NOTE! This get_declaration check applies only for pre frame pointer case
+        # the post frame pointer case should not apply
+        # need to somehow expose the context of evaluation
+        try:
+            declaration = self.computation.subroutine.get_declaration_by_option(False)
+
+            if declaration.deferred_expr is None:
+                raise TealInputError(
+                    "ABI return subroutine must have deferred_expr to be not-None."
+                )
+            if declaration.deferred_expr.type_of() != output.type_spec().storage_type():
+                raise TealInputError(
+                    f"ABI return subroutine deferred_expr is expected to be typed {output.type_spec().storage_type()}, "
+                    f"but has type {declaration.deferred_expr.type_of()}."
+                )
+        except Exception:
+            pass
```

Comment:
@ahangsu Request for clarification:  For scratch slot-based usage, is it _correct_ to ignore the raised exception?

I think I'm missing some intuition.  I'm imagining if an exception was thrown prior to the PR (with scratch slots), then it's _still_ an error to raise to the user.

Feel welcomed to suggest a live discussion if there's too much context to convey in writing.

Reply:
Refactored in https://github.com/algorand/pyteal/pull/606/commits/0f1fe4f384c0a829c3aa9188bd548f54b16e4821, for the only error throwing is running `get_declaration_by_option`, with no access to current compile option, such that compiling against scratch slot with a evaluated to frame pt body will error.

I refactored out the coarse scope of error catching in latest commit.